apps/sync/js/adapters/history.js

```diff
@@ -0,0 +1,283 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/**
+  History Data Adapter
+
+    The feature of History adapter is to retrieve History records from FxSync
+  Service and merge to Places DataStore which has the same functionality, and
+  the adapter is based on SyncEngine's design to handle 'update' and
+  'handleConflict' request. History Data Adapter provides Read-Only version
+  currently, and the basic ideas are included:
+  - Traverse all records from History Collection (HC).
+  - Add a new or existed place to Places DataStore (PDS).
+    - Convert the records from HC to PDS format.
+    - Merge a new record and the existed one then write to PDS.
+  - Remove an existed place when receiving a "deleted: true" record.
+**/
+
+'use strict';
+
+/* global
+  asyncStorage,
+  DataAdapters,
+  LazyLoader
+*/
+
+const HISTORY_COLLECTION_MTIME = 'collections::history::mtime';
+const SYNCTOID_PREFIX = 'SynctoId::';
+
+var HistoryHelper = (() => {
+  var placesStore;
+  function _ensureStore() {
+    if (placesStore) {
+      return Promise.resolve(placesStore);
+    }
+    return navigator.getDataStores('places').then(stores => {
+      placesStore = stores[0];
+      return Promise.resolve(placesStore);
+    });
+  }
+
+  function setSyncedCollectionMtime(mtime) {
+    return new Promise(resolve => {
+      LazyLoader.load(['shared/js/async_storage.js'], () => {
+        asyncStorage.setItem(HISTORY_COLLECTION_MTIME, mtime, resolve);
+      });
+    });
+  }
+
+  function getSyncedCollectionMtime() {
+    return new Promise(resolve => {
+      LazyLoader.load(['shared/js/async_storage.js'], () => {
+        asyncStorage.getItem(HISTORY_COLLECTION_MTIME, resolve);
+      });
+    });
+  }
+
+  function setSynctoIdMap(synctoId, dsId) {
+    return new Promise(resolve => {
+      LazyLoader.load(['shared/js/async_storage.js'], () => {
+        asyncStorage.setItem(SYNCTOID_PREFIX + synctoId, dsId, resolve);
+      });
+    });
+  }
+
+  function getSynctoIdMap(synctoId) {
+    return new Promise(resolve => {
+      LazyLoader.load(['shared/js/async_storage.js'], () => {
+        asyncStorage.getItem(SYNCTOID_PREFIX + synctoId, resolve);
+      });
+    });
+  }
+
+  function mergeRecordsToDataStore(existed, newPlace) {
+    if (existed.url !== newPlace.url) {
+      // The existed record has different url(id) with the new record.
+      throw new Error('Inconsistent records');
+    }
+    if (!existed.fxsyncId) {
+      existed.fxsyncId = newPlace.fxsyncId;
+    } else if(existed.fxsyncId !== newPlace.fxsyncId) {
+      // Two records have different fxsyncId but have the same url(id).
+      throw new Error('Inconsistent records');
+    }
+
+    existed.visits = existed.visits || [];
+    if (existed.visits.length === 0 && newPlace.title) {
+      existed.title = newPlace.title;
+    } else if (newPlace.visits[0] >= existed.visits[0]) {
+      existed.title = newPlace.title;
```

Comment:
Can you add a comment explaining this logic, please? Actually a comment explaining the whole merging logic would be great.